v2/internal/controllers/crd_networking_applicationgateway_test.go

```diff
@@ -0,0 +1,389 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package controllers_test
+
+import (
+	"log"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	network2020 "github.com/Azure/azure-service-operator/v2/api/network/v1api20201101"
+	network "github.com/Azure/azure-service-operator/v2/api/network/v1api20220701"
+	resources "github.com/Azure/azure-service-operator/v2/api/resources/v1api20200601"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/Azure/azure-service-operator/v2/internal/testcommon"
+	"github.com/Azure/azure-service-operator/v2/internal/util/to"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+func Test_Networking_ApplicationGateway_HTTPS_Listener_CRUD(t *testing.T) {
```

Comment:
This test needs to be recorded and have the resulting yaml file checked in.

Reply:
here I need your help to understand what is going wrong or in the way a run the test,  I'm running this command:
TIMEOUT=60m AZURE_TENANT_ID=16b3c013-d300-468d-ac64-7eda0820b6d3 AZURE_SUBSCRIPTION_ID=SUB_ID AZURE_CLIENT_ID=ClientID AZURE_CLIENT_SECRET=secrets TEST_FILTER=Test_Networking_ApplicationGateway_HTTPS_Listener_CRUD task controller:test-integration-envtest-record

In the subscription and from the logs everything is working, but the command end with this

[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:22Z] ApplicationGatewayController "msg"="Reconcile invoked" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" kind="&TypeMeta{Kind:ApplicationGateway,APIVersion:network.azure.com/v1api20220701storage,}" resourceVersion="1153" generation=2 uid="00e2641a-e857-4996-b9a1-4af068dbe6bb" ownerReferences=[]v1.OwnerReference{v1.OwnerReference{APIVersion:"resources.azure.com/v1api20200601storage", Kind:"ResourceGroup", Name:"asotest-rg-xofglu", UID:"e9fda82c-b286-49e1-89e3-d9ab6c010ba6", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}} creationTimestamp="2023-08-23 09:38:47 +0000 UTC" deletionTimestamp="2023-08-23 09:55:22 +0000 UTC" finalizers=[]string{"serviceoperator.azure.com/finalizer"} annotations=map[string]string{"serviceoperator.azure.com/resource-id":"/subscriptions/515ffc0b-a14f-4750-aef5-3db99394b8db/resourceGroups/asotest-rg-xofglu/providers/Microsoft.Network/applicationGateways/appgatewaytest"} conditions="[Condition [Ready], Status = \"True\", ObservedGeneration = 1, Severity = \"\", Reason = \"Succeeded\", Message = \"\", LastTransitionTime = \"2023-08-23 09:43:53 +0000 UTC\"]" owner="asotest-rg-xofglu, Group/Kind: resources.azure.com/ResourceGroup"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:22Z] ApplicationGatewayController "msg"="Determined Delete action" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="appgatewaytest" action="BeginDelete"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:22Z] ApplicationGatewayController "msg"="Starting delete of resource" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="appgatewaytest"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] PublicIPAddressController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="asotest-publicip-uvwecz" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="asotest-publicip-uvwecz"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] PublicIPAddressController "msg"="Delete succeeded, removing finalizer" name="asotest-publicip-uvwecz" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record] 2023/08/23 09:55:23 http: TLS handshake error from 127.0.0.1:56704: EOF
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:23Z] PublicIPAddressController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/asotest-publicip-uvwecz resource status: Operation cannot be fulfilled on publicipaddresses.network.azure.com "asotest-publicip-uvwecz": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/publicipaddresses/aso-test-networking-applicationgateway-https-listener-crud/asotest-publicip-uvwecz, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: e483dfb7-ef2b-4557-b595-a422bca35790, UID in object meta: " name="asotest-publicip-uvwecz" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] ApplicationGatewayController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="appgatewaytest"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] ApplicationGatewayController "msg"="Delete succeeded, removing finalizer" name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] VirtualNetworkController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="asotest-vn-fsazuy" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="asotest-vn-fsazuy"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:23Z] VirtualNetworkController "msg"="Delete succeeded, removing finalizer" name="asotest-vn-fsazuy" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:23Z] VirtualNetworkController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/asotest-vn-fsazuy resource status: Operation cannot be fulfilled on virtualnetworks.network.azure.com "asotest-vn-fsazuy": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/virtualnetworks/aso-test-networking-applicationgateway-https-listener-crud/asotest-vn-fsazuy, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 0d3de2d4-a2d4-427b-94b3-0a8b96a7f02c, UID in object meta: " name="asotest-vn-fsazuy" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:23Z] ApplicationGatewayController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/appgatewaytest resource status: Operation cannot be fulfilled on applicationgateways.network.azure.com "appgatewaytest": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/applicationgateways/aso-test-networking-applicationgateway-https-listener-crud/appgatewaytest, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 00e2641a-e857-4996-b9a1-4af068dbe6bb, UID in object meta: " name="appgatewaytest" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record] 2023/08/23 09:55:27 http: TLS handshake error from 127.0.0.1:40460: read tcp 127.0.0.1:34851->127.0.0.1:40460: read: connection reset by peer
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:27Z] VirtualNetworksSubnetController "msg"="Reconcile invoked" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" kind="&TypeMeta{Kind:VirtualNetworksSubnet,APIVersion:network.azure.com/v1api20201101storage,}" resourceVersion="1159" generation=2 uid="de424115-8da3-4fe6-bb8b-8ec8452e492e" ownerReferences=[]v1.OwnerReference{v1.OwnerReference{APIVersion:"network.azure.com/v1api20201101storage", Kind:"VirtualNetwork", Name:"asotest-vn-fsazuy", UID:"0d3de2d4-a2d4-427b-94b3-0a8b96a7f02c", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}} creationTimestamp="2023-08-23 09:38:32 +0000 UTC" deletionTimestamp="2023-08-23 09:55:27 +0000 UTC" finalizers=[]string{"serviceoperator.azure.com/finalizer"} annotations=map[string]string{"serviceoperator.azure.com/resource-id":"/subscriptions/515ffc0b-a14f-4750-aef5-3db99394b8db/resourceGroups/asotest-rg-xofglu/providers/Microsoft.Network/virtualNetworks/asotest-vn-fsazuy/subnets/AzureAppGtwSubnet"} conditions="[Condition [Ready], Status = \"True\", ObservedGeneration = 1, Severity = \"\", Reason = \"Succeeded\", Message = \"\", LastTransitionTime = \"2023-08-23 09:38:42 +0000 UTC\"]" owner="asotest-vn-fsazuy, Group/Kind: network.azure.com/VirtualNetwork"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:27Z] VirtualNetworksSubnetController "msg"="Determined Delete action" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="AzureAppGtwSubnet" action="BeginDelete"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:27Z] VirtualNetworksSubnetController "msg"="Starting delete of resource" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="AzureAppGtwSubnet"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:28Z] VirtualNetworksSubnetController "msg"="Successfully issued DELETE to Azure - resource was already gone" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud" azureName="AzureAppGtwSubnet"
[controller:test-integration-envtest-record]     logr.go:278: I2023-08-23T09:55:28Z] VirtualNetworksSubnetController "msg"="Delete succeeded, removing finalizer" name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     test_logger.go:160: I2023-08-23T09:55:28Z] VirtualNetworksSubnetController "msg"="Failed to commit object to etcd" "err"="updating aso-test-networking-applicationgateway-https-listener-crud/asotest-subnet-jsrwzs resource status: Operation cannot be fulfilled on virtualnetworkssubnets.network.azure.com "asotest-subnet-jsrwzs": StorageError: invalid object, Code: 4, Key: /registry/network.azure.com/virtualnetworkssubnets/aso-test-networking-applicationgateway-https-listener-crud/asotest-subnet-jsrwzs, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: de424115-8da3-4fe6-bb8b-8ec8452e492e, UID in object meta: " name="asotest-subnet-jsrwzs" namespace="aso-test-networking-applicationgateway-https-listener-crud"
[controller:test-integration-envtest-record]     kube_per_test_context.go:200: All resources deleted
[controller:test-integration-envtest-record] --- FAIL: Test_Networking_ApplicationGateway_HTTPS_Listener_CRUD (1050.23s)
[controller:test-integration-envtest-record] FAIL
[controller:test-integration-envtest-record] FAIL       github.com/Azure/azure-service-operator/v2/internal/controllers 1051.375s
[controller:test-integration-envtest-record] FAIL
task: Failed to run task "controller:test-integration-envtest-record": exit status 1

Reply:
We can take a  look and help with this bit.

Reply:
Reading through the log, I found something interesting.

The version first quoted log line seems to indicate the ApplicationGateway was successfully created:

```
I2023-08-23T09:55:22Z] ApplicationGatewayController 
Condition [Ready], Status = "True", ObservedGeneration = 1, Severity = "", Reason = "Succeeded", Message = "", LastTransitionTime = "2023-08-23 09:43:53 +0000 UTC"
```
Next, we see the test start to clean up:
```
I2023-08-23T09:55:22Z] ApplicationGatewayController
Starting delete of resource
```
But, the last log line is just 6 seconds later, which is way too fast for it to have successfully cleaned up by deleting everything:
```
I2023-08-23T09:55:28Z
```
I'll run the test locally to repro.